### PR TITLE
Grouped Gemm with output transpose

### PR DESCRIPTION
**Pull request by zjing14**

### CODE REVIEW

File: example/15_grouped_gemm/grouped_gemm_transpose_xdl_fp16.cpp [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        int M1 = S;
+        int N0 = NumHead;
+        int N1 = HeadDim;
+
+        int M = M0 * N1;
+        int N = N0 * N1;
+        int K = NumHead * HeadDim;
+
+        int StrideA = K;
+        int StrideB = K;
```
Comment (ltqin) [inline example/15_grouped_gemm/grouped_gemm_transpose_xdl_fp16.cpp:107]: chage to strideA strideB?

File: example/15_grouped_gemm/grouped_gemm_transpose_xdl_fp16.cpp [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+    std::vector<void*> p_c;
+
+    gemm_descs.reserve(group_count);
+
+    for(int i = 0; i < group_count; i++)
+    {
+        int B       = 16;
+        int S       = 64;
+        int NumHead = 16;
+        int HeadDim = 64;
```
Comment (ltqin) [inline example/15_grouped_gemm/grouped_gemm_transpose_xdl_fp16.cpp:95]: numberHead, headDim?

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+#include "device_base.hpp"
+
+namespace ck {
+namespace tensor_operation {
+namespace device {
+
+struct GemmDesc
+{
+    ck::index_t M, N, K;
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp:13]: M_ N_ K_

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    ck::index_t StrideA, StrideB, StrideC;
+};
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp:15]: stride_a_ stride_b_...?

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+struct GemmTransposeDesc
+{
+    ck::index_t M, N, K;
+    ck::index_t StrideA, StrideB;
+
+    ck::index_t M0, M1, N0, N1;
+    ck::index_t StrideM0, StrideM1, StrideN0, StrideN1;
+};
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp:24]: same above

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+struct DeviceGroupedGemm : public BaseOperator
+{
+    virtual std::unique_ptr<BaseArgument> MakeArgumentPointer(std::vector<const void*>& p_a,
+                                                              std::vector<const void*>& p_b,
+                                                              std::vector<void*>& p_c,
+                                                              std::vector<GemmDesc>& gemm_desc,
+                                                              AElementwiseOperation a_element_op,
+                                                              BElementwiseOperation b_element_op,
+                                                              CElementwiseOperation c_element_op,
+                                                              ck::index_t KBatch = 1) = 0;
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp:38]: k_batch?

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp [hunk 7]
```diff
@@ -0,0 +1,671 @@
+#ifndef DEVICE_GROUPED_GEMM_XDL_HPP
+#define DEVICE_GROUPED_GEMM_XDL_HPP
+
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp:3]: #pragma once
or 
#define DEVICE_GROUPED_GEMM_TRANSPOSE_XDL_HPP

File: library/include/ck/library/reference_tensor_operation/cpu/reference_gemm_transpose.hpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                const int n = n0 * arg.c_m0_m1_n0_n1_.mDesc.GetLengths()[3] + n1;
+
+                float v_acc = 0;
+
+                for(int k = 0; k < K; ++k)
+                {
+                    float v_a;
+                    float v_b;
+
+                    arg.a_element_op_(v_a, static_cast<const float>(arg.a_m_k_(m, k)));
```
Comment (ltqin) [inline library/include/ck/library/reference_tensor_operation/cpu/reference_gemm_transpose.hpp:66]: static_cast change to ck::type_convert

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp (hunk 6 above)
Comment (zjing14) [reply include/ck/tensor_operation/gpu/device/device_grouped_gemm.hpp:38]: It contains in the original code.

**Comment by zjing14** [comment]
@ltqin Fixed comments, please review the PR again and approve it.

**Comment by zjing14** [comment]
@ltqin Please review and approve the PR.

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp [hunk 9]
```diff
@@ -0,0 +1,677 @@
...
+        CGridDesc_M_N c_grid_desc_m_n_;
+
+        typename GridwiseGemm::CGridDesc_M0_N0_M1_N1_M2_M3_M4_N2
+            c_grid_desc_m0_n0_m1_n1_m2_m3_m4_n2_;
+
+        GroupedGemmBlock2CTileMap grouped_gemm_transpose_block_2_ctile_map_;
+
+        const ADataType* a_ptr;
+        const BDataType* b_ptr;
+        CDataType* c_ptr;
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp:352]: a_ptr_, b_ptr_, c_ptr_

**Review by ltqin — APPROVED** [review]

**Comment by zjing14** [comment]
New PR #288 contains all changes from this PR.

**Pull request closed without merging**